fem/kernels.hpp

```diff
@@ -26,6 +28,41 @@ namespace kernels
 namespace internal
 {
 
+/// @brief Resize an internal device memory suitable for a block of threads.
+/// When @a size is 0, the internal data is deleted and the nullptr is returned.
+template<int GRID, typename T = double>
+static T *DeviceMemSetSize(const int size)
```

Comment:
I like the simplification, less the hidden allocation inside the constructor.
Does this approach 'share' one single data for the different kernels?

Reply:
I don't see why the allocation is any more hidden in the constructor compared to in your function, I actually think it is more idiomatic to allocate in a constructor rather than in a function. I think the behavior of these classes is the same as your functions, mostly safer (checks are performed in debug) and cleaner (the logic between device vs shared memory is hidden from the kernel, and taking care of by the class). Shared memory is flushed between each kernel launch, device memory is not. The device memory is the same between kernels if the type `T` is the same *and* the `GRID` integer is the same, as in your case I believe.

I think that this design adds a few features, a `MemoryPool` could have more uses than just extra memory space for the thread blocks, in particular it can be a communication tool between blocks, or an optimization tool for reductions, e.g., dot product.

Reply:
At a first glance, 
```
const DeviceMatrix Bt(mp, d,q);
const DeviceMatrix Gt(mp, d,q);
```
`Bt` and `Gt` seem to overlap.

One feature that was targeted was to share this 'scratch' memory between the kernels, whereas this approach introduces nicely another memory pool feature for each kernel.

Reply:
If you prefer:
```c++
const DeviceMatrix Bt(mp.Alloc(d*q), d,q);
const DeviceMatrix Gt(mp.Alloc(d*q), d,q);
```
it also works. It's mostly a matter of taste.

Sharing "scratch" memory is also supported by this approach as far as I can tell. Can you give me a pseudo-code example of how you would use this "scratch" memory to exchange data between kernels in a way that is not supported by what I'm proposing? Or write a single line for each approach that results in different behaviors?

Reply:
Each `SharedMemoryPool` has a `GridMemoryPool` which will have its `static Memory<T> data;`, that is not the idea behind the shared device memory here.

Reply:
From what I can tell, there is two different ideas in your work:
- Having temporary memory spaces using device memory,
- Switching between shared memory and device memory temporary space for kernel usage.
The first point is done with either `MemoryPool` or `GridMemoryPool`, the second with `SharedMemoryPool`.

You also misunderstood the code I wrote, `SharedMemoryPool` has a shared memory pointer except in the case where `Size` is `0` where it is using device memory, as far as I know it is exactly the same behavior as the code you inlined in your kernels:
```c++
   const int sm_size = 2*d*q + 4*q*q;
   const int GRID = USE_SMEM ? 0 : 128;
   double *gmem = kernels::internal::DeviceMemSetSize<GRID>(sm_size); // device memory
   const int bid = MFEM_BLOCK_ID(x);
   constexpr int SM_SIZE = 2*D*Q + 4*Q*Q; // shared memory
   constexpr bool USE_SMEM = D > 0 && Q > 0;
   MFEM_SHARED double SMEM[USE_SMEM ? SM_SIZE : 1];
   double *sm = USE_SMEM ? SMEM : (gmem + sm_size*bid); // shared or device memory
```
is equivalent to:
```c++
   constexpr int SM_SIZE = 2*D*Q + 4*Q*Q;
   const int sm_size = 2*d*q + 4*q*q;
   const int GRID = 128;
   SharedMemoryPool<SM_SIZE,GRID> mp(sm_size); // shared or device memory
```

Reply:
This proposal introduces special need for Pool/Shared/Grid/Specialization/Allocation-through-Atomics.
Have you tested it ?

Reply:
Yes I have tested atomics...

Reply:
Have you tested the code you are proposing?
I'm having a hard time compiling it, but I might misunderstand something.

Reply:
I have not tested this code, this is a draft that I wrote directly in the comment (The `AtomicAdd` was meant to be a safety improvement). The reason why I think this would be useful is because even if officially we are trying to make kernels simpler to write, writing this kind of boiler-plate code that is in this PR goes in the direction of making kernel codes more complicated. If this approach is a path we want to follow, we should try to hide the complexity for users, all this feels very low level.